### ISSUE TITLE
Consistently failing test in nightly_win_xfail: rosbag2_tests.TestRecordEndToEnd/RecordFixture

### ISSUE THREAD

**Issue by clalancette** [ros2/rosbag2#1270]
## Description
The https://ci.ros2.org/view/nightly/job/nightly_win_xfail have consistently failing tests in the `RecordFixture` test since approximately forever (at least back into November 2022).

The most recent failures are here: https://ci.ros2.org/view/nightly/job/nightly_win_xfail/1095/#showFailuresLink

### Expected Behavior
Tests pass on the nightly builds.

### Actual Behavior
Full log is at https://ci.ros2.org/view/nightly/job/nightly_win_xfail/1095/testReport/ , but choosing one of the tests at random shows:

```
unknown file
C++ exception with description "Bag is not open. Call open() before setting filter." thrown in the test body.
```

## To Reproduce
** Steps to reproduce the behavior, e.g.
1. Run a job on Windows equivalent to the nightly windows xfail tests.
2. See the failure.

## System (please complete the following information)
 - OS: Windows 10
 - ROS 2 Distro: Rolling
 - Version: d926a617c12831b422f4eb364e5e886aec988461 (but this has been happening for a long time)

## Additional context
** Add any other context about the problem here **

**Comment by MichaelOrlov** [comment]
Preliminary RCA:
From the log:
```
1: -- run_test.py: invoking following command in 'C:\ci\ws\build\rosbag2_tests':
1:  - C:/ci/ws/build/rosbag2_tests/Release/test_rosbag2_record_end_to_end.exe --gtest_output=xml:C:/ci/ws/build/rosbag2_tests/test_results/rosbag2_tests/test_rosbag2_record_end_to_end.gtest.xml
1: Running main() from gmock_main.cc
1: [==========] Running 28 tests from 1 test suite.
1: [----------] Global test environment set-up.
1: [----------] 28 tests from TestRecordEndToEnd/RecordFixture
1: [ RUN      ] TestRecordEndToEnd/RecordFixture.record_end_to_end_test/0
1: stdin is not a terminal or console device. Keyboard handling disabled.[INFO] [1679987500.850727800] [rosbag2_recorder]: Press SPACE for pausing/resuming
1: [INFO] [1679987500.876319100] [rosbag2_storage]: Opened database 'C:\Users\ContainerAdministrator\AppData\Local\Temp\tmp_test_dir_a31248\record_end_to_end_test_0_sqlite3\record_end_to_end_test_0_sqlite3_0.db3' for READ_WRITE.
1: [INFO] [1679987500.879085300] [rosbag2_recorder]: Event publisher thread: Starting
1: [INFO] [1679987500.880136100] [rosbag2_recorder]: Listening for topics...
1: [INFO] [1679987501.001778100] [rosbag2_recorder]: Subscribed to topic '/test_topic'
1: [INFO] [1679987501.001961900] [rosbag2_recorder]: All requested topics are subscribed. Stopping discovery...
1: [WARN] [1679987501.383189300] [rosbag2_cpp]: No file paths were found in metadata.
1: unknown file: error: C++ exception with description "Bag is not open. Call open() before setting filter." thrown in the test body.
1: [  FAILED  ] TestRecordEndToEnd/RecordFixture.record_end_to_end_test/0, where GetParam() = "sqlite3" (1722 ms)
```
It looks like `reader->open(root_bag_path_.string());` silently fail here
https://github.com/ros2/rosbag2/blob/d926a617c12831b422f4eb364e5e886aec988461/rosbag2_tests/test/rosbag2_tests/record_fixture.hpp#L160
and it fails with log message `No file paths were found in metadata.` which is printed from here 
https://github.com/ros2/rosbag2/blob/d926a617c12831b422f4eb364e5e886aec988461/rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp#L95-L102
It might be an issue that we we are writing metadata on disk, the file exists but content is empty due to caching file in OS buffers. Anyway need to make further more deeper analysis why `metadata_.relative_file_paths` is empty.
@clalancette Am I understand correctly that it only fails on Windows build?

**Comment by clalancette** [comment]
> @clalancette Am I understand correctly that it only fails on Windows build?

Yes, I've only ever seen it in the nightly_win_xfail job.

**Comment by MichaelOrlov** [comment]
Update: It likely relates to how we are stopping recording on Windows.  
We are calling `stop_execution(..)` from here 
https://github.com/ros2/rosbag2/blob/d926a617c12831b422f4eb364e5e886aec988461/rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp#L147-L150
before trying to open metadata file.
And it has different implementation for Windoes and for Linux.
https://github.com/ros2/rosbag2/blob/d926a617c12831b422f4eb364e5e886aec988461/rosbag2_test_common/include/rosbag2_test_common/process_execution_helpers_windows.hpp#L108-L120
I have addressed some problems on Linux some time ago.  
It looks likes the problem in `stop_execution(..)` implementation for windows version. 
The problem is that on Unix we are sending `SIGINT` signal and handling it by default in `rclcpp` which is doing `rclcpp::shutdown();` and rosbag2 recorder gracefully exiting, although there are no `SIGINT` on Windows.  
I am not aware how this situation handling on Windows in ROS 2.  
@emersonknapp @clalancette any thoughts?

**Comment by clalancette** [comment]
> `rclcpp::shutdown();` and rosbag2 recorder gracefully exiting, although there are no `SIGINT` on Windows.
> I am not aware how this situation handling on Windows in ROS 2.

Yeah, I don't really know either.  I do know that Windows is very different in the signal handling.

> @emersonknapp @clalancette any thoughts?

Instead of using signals here, could we use another mechanism to shut-down the test cleanly?  Maybe a topic or a service call in the test program?

**Comment by MichaelOrlov** [comment]
@clalancette As regards to the:
> Instead of using signals here, could we use another mechanism to shut-down the test cleanly? Maybe a topic or a service call in the test program?

Actually not. This is subset of integration tests and we are trying to simulate how final user will be using rosbag2 recorder.  
Usually user running `ros2 bag record -a` in terminal and as soon as need to finish recording just pressing `CTRl+C` in the same terminal window. In test we are trying to reproduce exact same scenarios with different rosbag2 recorder options and cases and check that recording finishes correctly. On Unix `CTRL+C` mapped to the `SIGINT` - on Windows - some alternative which I am not very familiar with.
Sending message on dedicated topic or service call will defeat the purpose of the test.

**Comment by clalancette** [comment]
> Actually not. This is subset of integration tests and we are trying to simulate how final user will be using rosbag2 recorder. Usually user running `ros2 bag record -a` in terminal and as soon as need to finish recording just pressing `CTRl+C` in the same terminal window. In test we are trying to reproduce exact same scenarios with different rosbag2 recorder options and cases and check that recording finishes correctly. On Unix `CTRL+C` mapped to the `SIGINT` - on Windows - some alternative which I am not very familiar with. Sending message on dedicated topic or service call will defeat the purpose of the test.

OK, yeah, good point.  In that case, I don't really have much good advice here; I don't understand how signal handling works on Windows either.

**Comment by MichaelOrlov** [comment]
@clalancette I will try to make some investigation in this area but would like to postpone this task a bit for a month or two.
